src/core/geometry/qgslinestring.cpp

```diff
@@ -187,6 +187,90 @@ bool QgsLineString::isEmpty() const
   return mX.isEmpty();
 }
 
+QgsLineString *QgsLineString::asGridified( double hSpacing, double vSpacing, double dSpacing, double mSpacing,
+    double /*tolerance*/, SegmentationToleranceType /*toleranceType*/ ) const
+{
+  int length = numPoints();
+
+  if ( length <= 0 )
+    return nullptr;
+
+
+  QgsLineString *result;
+
+
+  // helper functions
+  auto roundVertex = [&]( QgsPoint & out, int i )
+  {
+    if ( hSpacing > 0 )
+      out.setX( round( mX.at( i ) / hSpacing ) * hSpacing );
+
+    if ( vSpacing > 0 )
+      out.setY( round( mY.at( i ) / vSpacing ) * vSpacing );
+
+    if ( dSpacing > 0 && QgsWkbTypes::hasZ( mWkbType ) )
+      out.setZ( round( mZ.at( i ) / dSpacing ) * dSpacing );
+
+    if ( mSpacing > 0 && QgsWkbTypes::hasM( mWkbType ) )
+      out.setM( round( mM.at( i ) / mSpacing ) * mSpacing );
+  };
+
+
+  auto append = [this, &result]( QgsPoint const & point )
+  {
+    result->mX.append( point.x() );
+
+    result->mY.append( point.y() );
+
+    if ( QgsWkbTypes::hasZ( mWkbType ) )
+      result->mZ.append( point.z() );
+
+    if ( QgsWkbTypes::hasM( mWkbType ) )
+      result->mM.append( point.m() );
+  };
+
+
+  auto isPointEqual = [&]( const QgsPoint & a, const QgsPoint & b )
+  {
+    return ( a.x() == b.x() )
```

Comment:
Probably safest to use `qgsDoubleNear( a.x(), b.x() )`, etc here.

Reply:
It is always called after snapping, so there are no small errors (and this way is faster) ^^